TRUSTED_OWNERS and TRUSTED_PATHS to protect against code injection via 'source'

jsmeix opened the pull request:
* Type: **Enhancement**

* Impact: **High**

* Reference to related issue (URL):
https://github.com/rear/rear/issues/3259

* How was this pull request tested?
Not yet tested

* Description of the changes in this pull request:

In default.conf describe the new config variables
TRUSTED_OWNERS and TRUSTED_PATHS
which are used to provide the new
general protection against code injection via 'source'
see https://github.com/rear/rear/issues/3259
and "Protecting Against Code Injections" in
https://relax-and-recover.org/documentation/security-architecture

jsmeix commented:
I tested "rear mkrescue" and "rear mkbackuponly".
Tomorrow I will test "rear recover"...

jsmeix commented:
I tested "rear recover".

So for me on a SLES15-SP6 test VM
(with the SLES15 default btrfs structure)
"rear mkrescue" plus "rear mkbackuponly"
and then on another same test VM "rear recover"
"just worked" - i.e. with the defaults for
TRUSTED_OWNERS and TRUSTED_PATHS.

I use this etc/rear/local.conf
(with the usual settings for the SLES15 default btrfs structure)
```
OUTPUT=ISO
BACKUP=NETFS
BACKUP_OPTIONS="nfsvers=3,nolock"
BACKUP_URL=nfs://192.168.178.66/nfs
REQUIRED_PROGS+=( snapper chattr lsattr )
COPY_AS_IS+=( /usr/lib/snapper/installation-helper /etc/snapper/config-templates/default )
POST_RECOVERY_SCRIPT=( 'if snapper --no-dbus -r $TARGET_FS_ROOT get-config | grep -q "^QGROUP.*[0-9]/[0-9]" ; then snapper --no-dbus -r $TARGET_FS_ROOT set-config QGROUP= ; snapper --no-dbus -r $TARGET_FS_ROOT setup-quota && echo snapper setup-quota done || echo snapper setup-quota failed ; else echo snapper setup-quota not used ; fi' )
SSH_ROOT_PASSWORD="rear"
USE_DHCLIENT="yes"
FIRMWARE_FILES=( 'no' )
MODULES=( 'loaded_modules' )
```
cf.
https://github.com/rear/rear/wiki/Test-Matrix-ReaR-2.8#sles-15-sp-6-with-default-btrfs-structure

jsmeix commented:
As far as I understand it
the currently failing checks
```
testing-farm:centos-stream-8-x86_64 — Failed to provision test environment.
Please check https://status.testing-farm.io for possible outages.
If the situation persists after retrying, please report an issue.

testing-farm:centos-stream-9-x86_64 — Failed to provision test environment.
Please check https://status.testing-farm.io for possible outages.
If the situation persists after retrying, please report an issue.
```
show on https://status.testing-farm.io the reason
```
Public Ranch Outage - bug in security groups cleanup.
...
This issue is not resolved yet
```
So it seems the reason for those currently failing checks
is outside of the ReaR environment - in particular the
reason is not the code changes in this pull request.

jsmeix commented:
@schlomo @gdha @pcahyna 
please have a look here - as time permits.

For the new script that combines
lib/_input-output-functions.sh and lib/framework-functions.sh
I chose the name _framework-setup-and-functions.sh
because the shorter name _framework-functions.sh
is misleading because in contrast to the other *functions.sh scripts
that special script does not only define framework functions
but also does some framework setup actions so it does both
framework setup and defining framework functions, in particular
input-output functions (from _input-output-functions.sh)
and sourcing functions (from framework-functions.sh).

jsmeix commented:
@rear/contributors 
as noone objected against my chosen name
`_framework-setup-and-functions.sh`
I will now change all ReaR scripts that mention
the old name `_input-output-functions.sh` (usuall in comments)
to show the new name `_framework-setup-and-functions.sh`

schlomo commented on usr/share/rear/conf/default.conf:
```diff
@@ -85,6 +87,49 @@
+# The empty TRUSTED_OWNERS array here means
+# that 'root' and the owner of usr/sbin/rear are trusted by default.
+# To add users to the default TRUSTED_OWNERS specify in etc/rear/local.conf
+# TRUSTED_OWNERS+=( 'user1' 'user2' )
```
```suggestion
# TRUSTED_OWNERS+=( user1 user2 )
```

jsmeix replied:
Ah - yes - "man useradd" tells it
```
Usernames must start with a lower case letter or an underscore,
followed by lower case letters, digits, underscores, or dashes.
They can end with a dollar sign. In regular expression terms:
[a-z_][a-z0-9_-]*[$]?
```
so user names cannot contain spaces.

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+#   VAR_DIR=/var/lib/rear
+#   LOG_DIR=/var/log/rear
+#
+# (A2)
```
What is the difference between A1 and A2?

jsmeix replied:
The difference is "on the original system" (A1)
versus "in the ReaR recovery system" (A2)
and in both cases REAR_DIR_PREFIX is empty
and the ReaR basic directories are same
on the original system and in the ReaR recovery system.

I listed "on the original system"
versus "in the ReaR recovery system"
sepatated also for case (A no Git checkout)
to have is same for case (A)
as it is then shown for case (B Git checkout)
where REAR_DIR_PREFIX and the ReaR basic directories are
different on the original system and in the ReaR recovery system.

So it is easier to compare case B1 with the matching A1
(both cases show how it looks on the original system)
versus case B2 with the matching A2
(both cases show how it looks in the ReaR recovery system)

jsmeix replied:
After re-reading the comments about cases A1 and A2 in
_framework-setup-and-functions.sh I see now that I can
simplify the cases A1 and A2 into one single case A
because I nowhere to handle one of the cases A1 or A2
separated from the other one but always both together
(in contrast to the cases B1 and B2 where I need to
handle B1 separated from B2 via different code).

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+    # and in the latter case we specify '/rear/prefix/' first in trusted_paths
+    # cf. https://github.com/rear/rear/pull/3379#issuecomment-2601767515
+    # cf. the REAR_DIR_PREFIX setting code in sbin/rear
+    # TODO: Perhaps test_path="$SHARE_DIR/lib/_framework-setup-and-functions.sh" is better than hardcoded '/usr/share/rear'?
```
```suggestion
```

jsmeix replied:
ok

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+# (i.e. when ReaR's files are installed in '/' so REAR_DIR_PREFIX is empty):
+#   REAR_DIR_PREFIX=
+#   SHARE_DIR=/usr/share/rear
+#   CONFIG_DIR=/etc/rear
```
In case it is not covered already: What about adding extra configs via `-C`? Shouldn't that path also be part of the `TRUSTED_PATHS`?

jsmeix replied:
See my below
https://github.com/rear/rear/pull/3424/files#r1992982825

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+# (i.e. when ReaR's files are installed in '/rear/prefix/' so REAR_DIR_PREFIX is non-empty '/rear/prefix'):
+#   REAR_DIR_PREFIX=/rear/prefix
+#   SHARE_DIR=/rear/prefix/usr/share/rear
+#   CONFIG_DIR=/rear/prefix/etc/rear
```
My usual workflow to test is to rsync my git checkout to the test VM and then call it like this:

```
./usr/sbin/rear -C /etc/rear/site.conf mkbackup
```

so that configs added by `-C` should be also considered here, but during recovery my "custom" `site.conf` is magically in the place where ReaR looks for it by default, so that I don't have to add the `-C` for the `rear recover` call :grin:

jsmeix replied:
How it behaves with additional config files
is currently only described in a comment in sbin/rear
```
# After the normal user configuration files (site.conf local.conf rescue.conf) were sourced
# set TRUSTED_OWNERS and TRUSTED_PATHS 'readonly' to avoid complicated automatisms for CONFIG_APPEND_FILES
# but this means when the user needs to source additional configuration files in non-default paths
# he must add his non-default paths to TRUSTED_PATHS in a normal user configuration file.
```

I think because you use a normal user configuration file
(site.conf local.conf rescue.conf) with -C
your case should "just work".

I will enhance the TRUSTED_OWNERS TRUSTED_PATHS
description in default.conf to also describe
how it behaves with additional config files.

jsmeix replied:
I fear that TRUSTED_OWNERS and TRUSTED_PATHS can currently
only be specified in the normal user configuration files
(site.conf local.conf rescue.conf)
but not in additional configuration files
causes the same kind of regression for the use case
```
# rear -C something_else_backup mkbackuponly
```
which had been reported for a different issue in
https://github.com/rear/rear/pull/3221#issuecomment-2129240645

So I think I may have to implement complicated things
for ONFIG_APPEND_FILES...

jsmeix replied:
I think it is sufficiently secure
to set TRUSTED_OWNERS and TRUSTED_PATHS 'readonly'
after all user configuration files were sourced
because all user configuration files are trusted.

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+    local file="$1"
+    local actual_path=""
+    local trusted_path=""
+    # Do not error out in 'readlink' when it is neither a regular file nor a link to a regular file
+    # but it is not trusted when it is neither a regular file nor a link to a regular file:
+    test -f "$file" || return 1
+    # Empty TRUSTED_PATHS means all regular file paths are blindly trusted,
+    # cf https://github.com/rear/rear/pull/3379#issuecomment-2633123866
+    test ${#TRUSTED_PATHS[@]} -eq 0 && return 0
```
```suggestion
    # Empty TRUSTED_PATHS means all regular file paths are blindly trusted,
    # cf https://github.com/rear/rear/pull/3379#issuecomment-2633123866
    test ${#TRUSTED_PATHS[@]} -eq 0 && return 0
    # Do not error out in 'readlink' when it is neither a regular file nor a link to a regular file
    # but it is not trusted when it is neither a regular file nor a link to a regular file:
    test -f "$file" || return 1
```

Isn't it empty by default? So we can check for the common scenario first

jsmeix replied:
Thank you!
This was clearly the wrong ordering
regardless that by default neither TRUSTED_OWNERS
nor TRUSTED_PATH are empty - they are empty in
default.conf but this means their defaults are set in
_framework-setup-and-functions.sh which is sourced
in sbin/rear before the user configs gets sourced.

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+    actual_path="$( readlink -e "$file" )" || Error "is_trusted_path(): 'readlink -e $file' failed"
+    for trusted_path in "${TRUSTED_PATHS[@]}" ; do
+        # Skip when trusted_path is empty (otherwise the [[ expression ]] would be falsely true):
+        test "$trusted_path" || continue
```
Maybe we better validate the `TRUSTED_PATHS` content once and error out if it contains blanks (or non-existent paths) to avoid checking it deep down each time and silently ignoring it?

How could there be an empty string in the list? Surely only as the result of manual misconfiguration

jsmeix replied:
Yes, this is to be fail-safe against misconfiguration
of the user.

I would allow non-existent paths because users may like
to use one same configuration on many systems,
for example something like
```
TRUSTED_PATHS+=( /opt/some_backup_software/ /opt/another_backup_software/ )
```
when they use more than one third-party backup software.

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+    for trusted_path in "${TRUSTED_PATHS[@]}" ; do
+        # Skip when trusted_path is empty (otherwise the [[ expression ]] would be falsely true):
+        test "$trusted_path" || continue
+        [[ $actual_path =~ ^$trusted_path ]] && return 0
```
Does that handle paths with spaces and special chars?

jsmeix replied:
I had tested it somewhat
but I fail to fully understand the quotation hell
and pattern matching description about `=~`
in the `[[ expression ]]` section in "man bash"
to be sure it will reall work in any case.

With GNU bash, version 4.4.23:
```
# TRUSTED_PATHS=( '/foo/' '/bar baz/' '/bar .* baz' )

# actual_path='/bar baz/file'

# (set -x ; for trusted_path in "${TRUSTED_PATHS[@]}" ; do [[ $actual_path =~ ^$trusted_path ]] && echo "matches '$trusted_path'" || echo "not matches '$trusted_path'" ; done )
+ for trusted_path in "${TRUSTED_PATHS[@]}"
+ [[ /bar baz/file =~ ^/foo/ ]]
+ echo 'not matches '\''/foo/'\'''
not matches '/foo/'
+ for trusted_path in "${TRUSTED_PATHS[@]}"
+ [[ /bar baz/file =~ ^/bar baz/ ]]
+ echo 'matches '\''/bar baz/'\'''
matches '/bar baz/'
+ for trusted_path in "${TRUSTED_PATHS[@]}"
+ [[ /bar baz/file =~ ^/bar .* baz ]]
+ echo 'not matches '\''/bar .* baz'\'''
not matches '/bar .* baz'

# (set -x ; for trusted_path in "${TRUSTED_PATHS[@]}" ; do [[ "$actual_path" =~ ^"$trusted_path" ]] && echo "matches '$trusted_path'" || echo "not matches '$trusted_path'" ; done )
+ for trusted_path in "${TRUSTED_PATHS[@]}"
+ [[ /bar baz/file =~ ^/foo/ ]]
+ echo 'not matches '\''/foo/'\'''
not matches '/foo/'
+ for trusted_path in "${TRUSTED_PATHS[@]}"
+ [[ /bar baz/file =~ ^/bar baz/ ]]
+ echo 'matches '\''/bar baz/'\'''
matches '/bar baz/'
+ for trusted_path in "${TRUSTED_PATHS[@]}"
+ [[ /bar baz/file =~ ^/bar \.\* baz ]]
+ echo 'not matches '\''/bar .* baz'\'''
not matches '/bar .* baz'
```

Because it works for me with
```
[[ $actual_path =~ ^$trusted_path ]]
```
I used that
but it also works for me with
```
[[ "$actual_path" =~ ^"$trusted_path" ]]
```
and the `set -x` output indicates this one is safer
so probably this one should be used
but I don't know what works more fail-safe?

schlomo commented on usr/sbin/rear:
```diff
@@ -50,23 +68,34 @@ function apply_bash_flags_and_options_commands () {
+# so we can reliably check such exceptional cases at any time to keep ReaR secure
+# (in particular we can also find where the undesirable form 'builtin .' is used):
+function source () {
+    echo -e "ERROR: BUG in $PRODUCT\nForbidden usage of 'source' in\n  source $*\nPlease report it at $BUG_REPORT_SITE" >&2
```
```suggestion
    echo -e "ERROR: BUG in $PRODUCT\nForbidden usage of Bash 'source' instead of ReaR 'Source' in\n  source $*\nPlease report it at $BUG_REPORT_SITE" >&2
```

jsmeix replied:
It is not about using ReaR's 'Source' function
because this function is only to source ReaR files
(ReaR scripts or ReaR config files).

It is about not using 'source'
until the source wrapper function in
_framework-setup-and-functions.sh is defined,
see my above
https://github.com/rear/rear/pull/3424/files#r1992949757

schlomo commented on usr/sbin/rear:
```diff
@@ -50,23 +68,34 @@ function apply_bash_flags_and_options_commands () {
+# and https://github.com/rear/rear/issues/3319#issuecomment-2360546784
+# and https://github.com/rear/rear/issues/3319#issuecomment-2363556217
+function . () {
+    echo -e "ERROR: BUG in $PRODUCT\nForbidden usage of '.' in\n  . $*\nPlease report it at $BUG_REPORT_SITE" >&2
```
```suggestion
    echo -e "ERROR: BUG in $PRODUCT\nForbidden usage of Bash '.' instead of ReaR 'Source' in\n  . $*\nPlease report it at $BUG_REPORT_SITE" >&2
```

jsmeix replied:
It is not about using ReaR's 'Source' function
because this function is only to source ReaR files
(ReaR scripts or ReaR config files).

I will change the text to
```
ERROR: BUG in $PRODUCT\nForbidden usage of '.' instead of 'source' in\n  . $*\nPlease report it at $BUG_REPORT_SITE
```

schlomo commented on usr/sbin/rear:
```diff
@@ -50,23 +68,34 @@ function apply_bash_flags_and_options_commands () {
+# The point is that we can reliably find all code places where 'builtin' is used
+# so we can reliably check such exceptional cases at any time to keep ReaR secure
+# (in particular we can also find where the undesirable form 'builtin .' is used):
+function source () {
```
As we replace this `source` implementation later on with the wrapper, what does this definition here protect? Can it be that it only protects the code between this definition and the redefinition later on? So specifically this definition protects the code in the `rear` main script between line 95 and line 554 where we load the `_framework-setup-and-functions.sh` which in turns redefines `source` to be the wrapper.

I'm wondering if this is maybe overkill as I expect us to trust ourselves to not wrongly use `source` here.

jsmeix replied:
Yes, it is only there to protect **all code**
until the source wrapper function in
_framework-setup-and-functions.sh is defined.

So it protects not only code in sbin/rear but also
code in default.conf which we 'builtin source' in sbin/rear
and also all code in _framework-setup-and-functions.sh
until the source wrapper function is defined therein
which are more than 1000 lines of code therein
until the source wrapper function is defined.

Yes, I don't trust ourselves or any contributor
to never use source by accident in the next 10 years
but then this BUG error shows up immediately
so the one who uses source by accident
gets notified already when he tests his things.

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,324 @@ function ProgressInfo () {
+    local file="$1"
+    local owner_name=""
+    local trusted_owner=""
+    # Do not error out in 'stat' when it is neither a regular file nor a link to a regular file
+    # but it is not trusted when it is neither a regular file nor a link to a regular file:
+    test -f "$file" || return 1
+    # Empty TRUSTED_OWNERS means all regular file owners are blindly trusted,
+    # cf https://github.com/rear/rear/pull/3379#issuecomment-2633123866
+    test ${#TRUSTED_OWNERS[@]} -eq 0 && return 0
```
```suggestion
    # Empty TRUSTED_OWNERS means all regular file owners are blindly trusted,
    # cf https://github.com/rear/rear/pull/3379#issuecomment-2633123866
    test ${#TRUSTED_OWNERS[@]} -eq 0 && return 0
    # Do not error out in 'stat' when it is neither a regular file nor a link to a regular file
    # but it is not trusted when it is neither a regular file nor a link to a regular file:
    test -f "$file" || return 1
```

test first for common case

jsmeix replied:
Thank you!
This was clearly the wrong ordering.

schlomo reviewed (approved):
Thanks a lot @jsmeix for reworking this. I carefully read through the code and added some suggestions and questions.

In general this looks already very good to me and it makes sense the way it is written now.

Maybe you can use quoting only where actually needed (e.g. no need to quote `'/lib'` as it is a plain string without spaces?

jsmeix commented:
@rear/contributors 
I need your help regarding how to properly use
```
[[ ... =~ ... ]]
```
and in particular regarding my current usage
in _framework-setup-and-functions.sh
```
actual_path="$( readlink -e "$file" )" || Error "is_trusted_path(): 'readlink -e $file' failed"
for trusted_path in "${TRUSTED_PATHS[@]}" ; do
    # Skip when trusted_path is empty (otherwise the [[ expression ]] would be falsely true):
    test "$trusted_path" || continue
    [[ $actual_path =~ ^$trusted_path ]] && return 0
done
```
cf. https://github.com/rear/rear/pull/3424#discussion_r1992872215
and https://github.com/rear/rear/pull/3424#discussion_r1993048851

This is how `[[ ... =~ ... ]]` is currently used in ReaR
```
# find usr/sbin/rear usr/share/rear/ -type f | xargs grep -oh '\[\[.*=~.*\]\]'

[[ $latest_tar_restore_file =~ .*/.* ]]
[[ $latest_tar_restore_file =~ .*/.* ]]
[[ "$BLOCKCLONE_ALLOW_MOUNTED" =~ ^[nN0] ]]
[[ $ID_NEW =~ ^dm- ]]
[[ $1 =~ /dev/mapper/ ]]
[[ $fs_spec =~ ^/dev/disk/by-id/ ]]
[[ $DEV_NAME =~ ^dm- ]]
[[ $1 =~ /dev/disk/by-id ]]
[[ $major_version =~ ^[0-9]+$ ]]
[[ $BORGBACKUP_ARCHIVE_PREFIX =~ [^a-zA-Z0-9] ]]
[[ "\$REPLY" =~ ^[Yy1] ]]
[[ "$UDEV_UID_LED" =~ ^[yY1] ]]
[[ "$UDEV_UID_LED" =~ ^[yY1] ]]
[[ "$DEVPATH" && "$UDEV_SUSPEND" =~ ^[yY1] ]]
[[ "$UDEV_BEEP" =~ ^[yY1] ]]
[[ $disk_name =~ ^dm- ]]
[[ "$name" =~ ^mapper/ ]]
[[ "$name" =~ ^dm- ]]
[[ ${device_name: -1} =~ [0-9] ]]
[[ "$test_ip" =~ $ip_pattern ]]
[[ $length =~ [^0-9] ]]
[[ $BACKUP_INTEGRITY_CHECK =~ ^[yY1] && "$(basename ${BACKUP_PROG})" = "tar" ]]
[[ "$BLOCKCLONE_ALLOW_MOUNTED" =~ ^[nN0] ]]
```

Currently I use totally unquoted
```
[[ $actual_path =~ ^$trusted_path ]]
```
which seems to work rather well for me, see
https://github.com/rear/rear/pull/3424#discussion_r1993048851

But probably I should better use
```
[[ "$actual_path" =~ ^$trusted_path ]]
```
or
```
[[ "$actual_path" =~ ^"$trusted_path" ]]
```
which also work for me, see
https://github.com/rear/rear/pull/3424#discussion_r1993048851

jsmeix commented:
@rear/contributors 
tomorrow I will test it again with all my recent changes
and if it still works well for my test case, cf.
https://github.com/rear/rear/pull/3424#issuecomment-2716931938
and when there are no objections
I would like to merge it next Monday afternoon

schlomo commented on usr/share/rear/lib/_framework-setup-and-functions.sh:
```diff
@@ -1507,3 +1507,316 @@ function ProgressInfo () {
+# ReaR basic directories cases:
+#
+# (A)
+# Normal case (i.e. no Git checkout) on the original system:
```
```suggestion
# Normal case (i.e. no Git checkout) on the original or recovery system:
```

jsmeix replied:
Not exactly what I like to tell so I did
https://github.com/rear/rear/pull/3424/commits/4e3f93ccdc7c76798969cb8a952bcbbfe58b8881
to improve the TRUSTED_PATHS description comment
of the "normal case (A)"

jsmeix commented:
Because "All checks have passed" and because
it also works for me (as far as I tested it)
I will merge it as is right now.

I must postpone the "cosmetic" part in
https://github.com/rear/rear/pull/3424#issuecomment-2720074872
```
... change all ReaR scripts that mention
the old name _input-output-functions.sh (usually in comments)
to show the new name _framework-setup-and-functions.sh
```
to a subsequent pull request (hopefully rather soon - but
I cannot promise it).

jsmeix commented:
Via https://github.com/rear/rear/pull/3379
where this pull request here is based on
I introduced a regression, see
https://github.com/rear/rear/pull/3379#issuecomment-2735731302
which I fixed now via
https://github.com/rear/rear/commit/d25816643715ee7c7fb4477be6bd8993663ca268

jsmeix commented:
While I did
https://github.com/rear/rear/pull/3424#discussion_r2009556786
I found another regression:
```
# usr/sbin/rear -s mkrescue
Relax-and-Recover 2.9 / 2025-01-31
Running rear mkrescue (PID 23165 date 2025-03-24 07:55:42)
Using log file: /root/rear.github.master/var/log/rear/rear-localhost.log
Simulation mode activated, Relax-and-Recover base directory: /root/rear.github.master/usr/share/rear
Source lib/array-functions.sh
...
Source lib/mkrescue-workflow.sh
...
Source lib/write-protect-functions.sh
Source conf/Linux-i386.conf
Source conf/GNU/Linux.conf
Source init/default/001_verify_config_arrays.sh
Source init/default/002_check_rear_recover_mode.sh
Source init/default/010_EFISTUB_check.sh
Source init/default/010_set_drlm_env.sh
Source init/default/030_update_recovery_system.sh
Source init/default/100_check_stale_nfs_mounts.sh
Source init/default/950_check_missing_programs.sh
Source init/default/998_dump_variables.sh
ERROR: The specified command 'mkrescue' does not exist!
Exiting rear mkrescue (PID 23165) and its descendant processes ...
Running exit tasks
```

This is because of
https://github.com/rear/rear/pull/3424/commits/7df7dc6499b7703b98d7293fbbf5de4d514a29ac
which results that also `lib/*-workflow.sh` scripts
get sourced by the Source function in simulation mode
so that e.g. in `lib/mkrescue-workflow.sh`
the WORKFLOW_mkrescue() function is not set.

jsmeix commented on usr/sbin/rear:
```diff
@@ -519,22 +551,34 @@ if ! test "$can_run_simultaneously" ; then
+# Additionally TRUSTED_OWNERS and TRUSTED_PATHS are set to default values
+# but TRUSTED_OWNERS and TRUSTED_PATHS are not yet set 'readonly' because this is done
+# after the normal user configuration files (site.conf local.conf rescue.conf) were sourced
+# so that the user can specify what he needs, e.g. for whatever third-party (backup) software
```
https://github.com/rear/rear/pull/3435
deletes 800_enforce_usb_output.sh which results
deletion of the whole usr/share/rear/build/USB directory
because 800_enforce_usb_output.sh is the only file therein.

pcahyna replied:
@jsmeix

> In this example sourcing etc/rear/local.conf again
(i.e. sourcing its copy $ROOTFS_DIR/etc/rear/local.conf)
at a later stage during "rear mkrescue" could cause
arbitrary unexpected misbehaviour of ReaR,
e.g. assume in etc/rear/local.conf there is
>```
> ARR+=( this that )
>```
> then after sourcing etc/rear/local.conf a second time
those array contais 'this' and 'that' each two times
so all code that deals with that array elements would
have to be idempotent to behave generically correctly
(by luck things may often even work somehow).

no, I believe this is actually "fine" (for a generous definition of "fine", of course). The sourcing happens in a subshell (`$(...)` spawns a subshell) and then one echoes the variable value. All the unfortunate non-idempotent modifications of variables are then blissfully forgotten. Recall that we discussed sourcing in a subshell and echoing the value as a possible "safe" (read: safe against unintended modification of the calling script's global context, not safe against execution of arbitrary code - note that local.conf should be considered trusted, so we care only about the first meaning of "safe" here) way to read values from shell-style configuration files some time ago.

The code has other problems of course: note that one can't be even sure that `OUTPUT` is set in `local.conf` - what if it is set in `site.conf`? (These two should be considered equivalent for all purposes, right?) and the purpose of the whole stuff is not well-defined, so I will be glad to see it disappear.

jsmeix replied:
@pcahyna 
of course you are right!

Those tiny little parentheses make such a difference ;-)

jsmeix replied:
Fixed via https://github.com/rear/rear/pull/3435

jsmeix commented:
https://github.com/rear/rear/pull/3436
intends to fix the above
https://github.com/rear/rear/pull/3424#issuecomment-2747079736

jsmeix commented:
The regression described in the initial comment here
```
There will be regressions when third-party scripts
use '.' for sourcing because currently sourcing via '.'
is always forbidden in ReaR.
```
see also
https://github.com/rear/rear/pull/3434#issuecomment-2742598092
should get fixed via
https://github.com/rear/rear/pull/3437
